### ISSUE TITLE
row/colMaxs, row/colMins, and row/colRanges crash R on a complex matrix

### ISSUE THREAD

**Issue by hpages** [HenrikBengtsson/matrixStats#95]
Hi Henrik,

I get the following segfault on my Ubuntu laptop with R 3.4 or R 3.3.1:
```
> library(matrixStats)
> m <- matrix(1:15, ncol=3)
> rowMaxs(m + 1i)

 *** caught segfault ***
address 0x3, cause 'memory not mapped'

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace
Selection: 
```
Maybe the type could be checked early with an error like:
```
> apply(m + 1i, 1, max)
Error in FUN(newX[, i], ...) : invalid 'type' (complex) of argument
```
Thanks,
H.
```
> sessionInfo()
R Under development (unstable) (2017-02-23 r72250)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.5 LTS

Matrix products: default
BLAS: /home/hpages/R/R-3.4.r72250/lib/libRblas.so
LAPACK: /home/hpages/R/R-3.4.r72250/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] matrixStats_0.52.1

loaded via a namespace (and not attached):
[1] compiler_3.4.0
```

**Comment by hpages** [comment]
FWIW I also get these segfaults on raw and character matrices. Thanks! H.

**Comment by HenrikBengtsson** [comment]
Thanks for the report. I can reproduce this. (My initial reaction is surprise, because we've got internal validation of data types, but on the other hand I also know there's almost always something overlooked).  This one gets the highest priority on this project.

**Comment by HenrikBengtsson** [comment]
This is now fixed in the `develop` branch (commit 899d626), e.g.
```r
> library("matrixStats")
>  m <- matrix(1:15, ncol=3)
>  rowMaxs(m + 1i)
Error in rowMaxs(m + (0+1i)) : 
  Argument 'x' must be of type logical, integer or numeric, not 'complex'.
```
I don't know how long this bug has been there, or when it was introduced, but in the previous type validation we somehow forgot to check for non-supported data types.

Thanks again for the report.  I'm curious, did you stumble upon this yourself, or did you discover it part of a bigger troubleshooting task, or because someone else reported a problem?

**Comment by hpages** [comment]
Thanks for the quick fix! I stumbled over this while implementing `row/colMaxs`, `row/colMins`, and `row/colRanges` methods for DelayedMatrix objects (DelayedArray package in Bioconductor). In order to mimic exactly how these functions behave on ordinary matrices, I had to pay special attention to the type of input they support and to understand the relationship between input and output type (normally `numeric`, except when input type is `integer` or `logical` and all the values in the output are finite, in which case it's `integer`).
